src/heartbeats.cpp

```diff
@@ -452,330 +299,120 @@ MegaBackupMonitor::MegaBackupMonitor(MegaClient *client)
-{
-    bool needsAdding = true;
-
-    // remove the tag from the set of in-flight sync registrations
-    mPendingSyncPuts.erase(syncTag);
-
-    if (ISUNDEF(backupId))
-    {
-        LOG_warn << "Received invalid id for sync with tag: " << syncTag;
-        needsAdding = false;
-    }
-
-    if (!ISUNDEF(backupId) && mHeartBeatedSyncs.find(syncTag) != mHeartBeatedSyncs.end())
-    {
-        mHeartBeatedSyncs[syncTag]->setBackupId(backupId);
-        needsAdding = false;
-    }
 
-    if (needsAdding)
-    {
-        for (const auto &hBPair : mHeartBeatedSyncs)
-        {
-            if (hBPair.second->backupId() == backupId)
-            {
-                needsAdding = false;
-                break;
-            }
-        }
-
-        if (needsAdding)
-        {
-            //create new HeartBeatSyncInfo
-            mHeartBeatedSyncs[syncTag] = std::make_shared<HeartBeatSyncInfo>(syncTag, backupId);
-        }
-    }
-
-    // store the id in the sync configuration
-    if (!ISUNDEF(backupId))
-    {
-        mClient->updateSyncBackupId(syncTag, backupId);
-    }
-
-    // handle pending updates
-    const auto &pendingSyncPair = mPendingSyncUpdates.find(syncTag);
-    if (pendingSyncPair != mPendingSyncUpdates.end())
-    {
-        if (!ISUNDEF(backupId))
-        {
-            MegaBackupInfo &info = *pendingSyncPair->second.get();
-            info.setBackupId(backupId);
-            updateBackupInfo(info);
-        }
-        else
+void MegaBackupMonitor::digestPutResult(handle backupId, UnifiedSync* syncPtr)
+{
+#ifdef ENABLE_SYNC
+    mClient->syncs.forEachUnifiedSync([&](UnifiedSync& us){
+        if (&us == syncPtr)
         {
-            LOG_warn << "discarding heartbeat update for pending sync: no valid id received for sync: " << pendingSyncPair->second->localFolder();
-
+            us.mConfig.setBackupId(backupId);
+            mClient->syncs.saveSyncConfig(us.mConfig);
         }
-        mPendingSyncUpdates.erase(pendingSyncPair);
-    }
-}
+    });
 #endif
-
-void MegaBackupMonitor::digestPutResult(handle backupId)
-{
-    if (mPendingBackupPutCallbacks.empty())
-    {
-        return;
-    }
-    // get the tag from queue of pending puts
-    auto putResultCallback = mPendingBackupPutCallbacks.front();
-    mPendingBackupPutCallbacks.pop_front();
-    //call corresponding callback
-    putResultCallback(backupId);
 }
 
-void MegaBackupMonitor::updateBackupInfo(const MegaBackupInfo &info)
+void MegaBackupMonitor::updateBackupInfo(handle backupId, const MegaBackupInfo &info)
 {
     string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", info.localFolder()) );
     string deviceIdHash = mClient->getDeviceidHash();
 
-    mClient->reqs.add(new CommandBackupPut(mClient, info.backupId(), info.type(), info.megaHandle(),
+    mClient->reqs.add(new CommandBackupPut(mClient,
+                                           info.type(),
+                                           info.backupName(),
```

Comment:
The backup name is not updated this way, but in the corresponding user's attribute.
```suggestion
```

Reply:
It looks like the command is saving that name (but not sending it) and in CommandBackupPut::procresult() it will do something complicated with it, that maybe adds it to client->mPendingBackupNames to be updated in the user attribute.  But, this scheme was not introduced by this PR.

Reply:
Yes. The backup name is a param of the command to register a new backup, and once we have a backup id, we add an item to the TLV of the user's attribute storing backup names (`backupd_id:backup_name` records). However, since multiple new backups can be registered in a single HTTP request (with multiple `sp` commands in parallel), and whenever any of them finishes there could be a `upv` in flight, we need to enqueue the updates and apply them only with a fresh version of the attribute (to avoid dirty writes).

Anyway, the comment was only relative to the fact that this call is for an update of the backup via `sp`, and the name is not part of the ctor for updates (only for registrations). So it needs to be deleted, as you already did :smiley: